### DISCUSSION TITLE
How to use Prisma Middleware?

### DISCUSSION THREAD

**Discussion by Kiodok** [blitz-js/blitz#3339] [Q&A] [answered]
Hello dear Blitz Community,

I'm working on a project right now, that will be released in a few month.
Therefore I need to encrypt some sensitive data in the database and also want to decrypt it on demand.

https://www.prisma.io/docs/concepts/components/prisma-client/middleware
The Prisma Client can use Middleware and I found an package that would do the job:
https://github.com/47ng/prisma-field-encryption

I tried a few hours to implement this package as Middleware to Prisma, but nothing worked. I think it's because Blitz enhanced the Prisma Client, but is there a way to integrate this Middleware?

Best regards!

**Comment by beerose** [comment, marked as answer]
Hey @Kiodok!

Just to confirm — `client.$use(fieldEncryptionMiddleware())` (where client is the client defined in `db/index.ts`) doesn't work? I know that some people successfully added a middleware with something like this:

```
import { enhancePrisma } from "blitz"
import { PrismaClient } from "@prisma/client"

const EnhancedPrisma = enhancePrisma(PrismaClient)
const prisma = new EnhancedPrisma()

prisma.$use(async (params, next) => {
  // Apply any middleware for prisma here.
  console.log(params.action, params.model)
  const result = await next(params)

  return result
})

export * from "@prisma/client"
export default prisma
```

**Reply by Kiodok** [reply]
Hey @beerose :)
Thanks for your help so far, encryption is now working as expected.
For testing I encrypted an email and in the Application it is automatically decrypted.

But when I tried to update the user, I got the following error:
Any ideas?

`12:54:35.778 ERROR updateUser() 
 Error 
Invalid `db__WEBPACK_IMPORTED_MODULE_2__.default.user.update()` invocation in
F:\Blitz Projects\streamconnect\.next\server\pages\api\rpc\updateUser.js:58:67

   55
   56 // TODO: in multi-tenant app, you must add validation to ensure correct tenant
   57 //const user = await db.user.update({ where: { id }, data })
→  58 const user = await db__WEBPACK_IMPORTED_MODULE_2__.default.user.update(
  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState(E54000), message: "index row requires 85848 bytes, maximum size is 8191", detail: None, hint: None, position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("indextuple.c"), line: Some(185), routine: Some("index_form_tuple") }) }) })
details:
{
  clientVersion: '3.14.0'
}
error stack:
• index.js:45409 request
    node_modules\@prisma\client\runtime\index.js:45409:15

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22

• index.js:20 fieldEncryptionMiddleware
    node_modules\prisma-field-encryption\dist\index.js:20:22`

**Reply by Kiodok** [reply]
Found the problem, I just needed to add these _previewFeatures = ["interactiveTransactions", "fullTextSearch", "fullTextIndex"]_ to the generator client in the schema.prisma :D
Thanks a lot for your help! :)
